### DISCUSSION TITLE
Initial ideas for framework's modeling approach

### DISCUSSION THREAD

**Discussion by sergefdrv** [replica-io/replica-io#49] [Ideas]
This is a discussion for gathering initial ideas concerning the framework's modeling approach as described in #47.

- [Summary of the thoughts related to modeling of distributed systems from the most abstract level](https://github.com/orgs/replica-io/discussions/49?sort=old#discussioncomment-12501075)

**Comment by sergefdrv** [comment]
Perhaps, we should start thinking about the framework's modeling approach from the most abstract level: the distributed system as a whole. At that level of abstraction, the distributed system can be represented as a single entity. Obviously, it doesn't make sense to consider it in total isolation, for it must exist within and somehow interact with the rest of the world.

Here it already starts getting interesting and we can ponder on the following questions:
 - What could be considered a part of the external world and what could belong to the internals of the distributed system?
 - How could we represent the external world w.r.t. the distributed system?
 - How could we think about the state of the distributed system and the state of the external world?
 - What is time w.r.t. the distributed system and the external world and how could it manifests itself in the system model?
 - How could we represent non-determinism in the system model?
 - How could we model the interaction of the distributed system with the external world?
 - How could we think about failures at this level of abstraction?

**Reply by sergefdrv** [reply]
Summary of the thoughts related to modeling of distributed systems from the most abstract level:
1. [Delimiting the system boundary based on the system's purpose (function)](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10588723), i.e. everything required to fulfill the purpose, and nothing more, is considered internal to the system, making it sufficient and complete.
2. [System's purpose as reducing uncertainty (non-determinism) imposed by the environment](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10738417).
3. [Time as causally consistent partial ordering of non-deterministic changes and system dynamics as resolving non-determinism imposed from the environment](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10661687).
4. [Bounded amount of determinism that can be exposed by a finite system resolving non-determinism imposed on it from the environment](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10881169).
5. [System state as information necessary and sufficient for preserving the system's causal consistency](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10729558).
6. [Interaction as causal dependencies between the interacting systems and redistribution of non-determinism from outside](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10738392).
7. [Interaction implies sharing common state](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10738566).
8. [Failure due to exceeding amount of non-determinism and leakage of the excessive non-determinism](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10740368).
9. [Using the concepts of "angelic" and "demonic" non-determinism](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10767510).
10. [Interaction as communicating partial information about causal dependencies between past events](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10880881).
11. [Initial relationships (connections, links) with the environment characterizing the system and enabling interactions; enabling new interactions by establishing connections dynamically as a result of prior interaction](https://github.com/orgs/replica-io/discussions/49#discussioncomment-11214071).
12. [New interaction possibly eliminating previously enabled interactions](https://github.com/orgs/replica-io/discussions/49#discussioncomment-11214526), e.g. [using single-shot interactions](https://github.com/orgs/replica-io/discussions/49#discussioncomment-12416989).
13. [Evolution of an open system as a reversible process of monotonically acquiring information from and monotonically returning information back to its environment; system state as initial and acquired information subtracted discarded information; constrains linking acquiring and discarding of specific information; defining systems and their components with acquiring and discarding of information made explicit](https://github.com/orgs/replica-io/discussions/49#discussioncomment-12273776).
14. [Using generalized controlled-swap gate for modeling races in a conceptually deterministic and reversible manner; optionally optimizing out reversibility and determinism](https://github.com/orgs/replica-io/discussions/49#discussioncomment-12425172).
15. [Modeling distributed and concurrent systems as a network of deterministic components interacting with each other and the environment through concurrent, pairwise connected lines of interaction, with complementary input and output sides, carrying plain data items of definite size (in binary representation); interaction as exchange of information between data items at different lines of interaction creating causal dependencies; unconnected inputs/outputs as external interface](https://github.com/orgs/replica-io/discussions/49#discussioncomment-12436415).
16. [Time as causal dependencies between inputs/outputs and timestamp values obeying certain assumptions](https://github.com/orgs/replica-io/discussions/49#discussioncomment-12491081).
17. [Component's failure as a violation (or weakening) of guarantees about its outputs; fault tolerance from guarantees being stronger than requirements](https://github.com/orgs/replica-io/discussions/49#discussioncomment-12492675).

**Comment by anchalshivank** [comment]
### What could be considered a part of the external world and what could belong to the internals of the distributed system?
**Internals of the Distributed System**
1. **Nodes** -> part of distributed system that has its own local state , computation  and communication abilities
2. **Communication channels** or the networks that connect the nodes

**External World**
1. **Clients** or **users** who interact with the system 
2. **External services** like **databases**, **apis** or other **distributed system** 
3. **Environment** in which the distributed system runs , including the **physical network** , **hardware** or **cloud infrastructure**

**Reply by sergefdrv** [reply]
What is the criteria for distinguishing what belongs to the internals and what is external?

**Reply by anchalshivank** [reply]
The key distinction between the **internals** of the distributed system and the **external world** is largely based on control and responsibility:

- **Internal** components are those that we can control , configure and define. We are responsible for their behavior and operation (nodes, algorithms and communication logic).
- **External** components are those outside our direct control, but we must design the system to cope with them (clients, external services, infrastructure).

**Reply by sergefdrv** [reply]
I think we are getting closer to the essence of your point, but there were new entities introduced into the discourse. Who are "we", who is responsible for the behavior of the distributed system being designed? Are we a part of the distributed system? Where do our responsibilities come from that determine what is distributed system's internal and what's not?

**Reply by anchalshivank** [reply]
**1. Who are "We"?**
In this context, **"we"** refers to the designers or developers responsible for building, configuring, and maintaining the distributed system.

**2. Are we part of the distributed system?**
**No**, we are not part of the system itself. We are not nodes or part of the infrastructure. Instead, we are external actors responsible for the system’s design and management.

**3. Where do our responsibilities come from that determine what is internal or external to the distributed system?**
Our responsibilities in defining what is internal or external to the distributed system stem from the **degree of control** over the system component.

**Internal Responsibilities:**
We are responsible for everything we control, design, configure, and directly manage. This includes components like nodes, communication protocols, and algorithms.
These components are considered internal because we shape their behavior and operation directly.
**External Responsibilities:**
We are responsible for ensuring the system interacts properly with components outside our control, such as clients, external services, and infrastructure.
These components are external because they exist independently, but we need to design our system to adapt and interact with them effectively.

**Reply by sergefdrv** [reply]
Do you mean that we are more or less free, and in fact that is one of our responsibilities, to define what is internal and what is external w.r.t. the distributed system we model and design, expect for the components that we (or rather the distributed system that we define) cannot control? If we decide to include something as a part of the system then doesn't it necessarily mean that the encompassing system takes control of that something in our model due to the fact that it is now considered a part of the system?

**Reply by sergefdrv** [reply]
Perhaps, the point of view on the distinction between internals and externals based on control and responsibility is somewhat similar to the point of view in terms of uncertainty and non-determinism (see [here](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10738417) and [here](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10661687)).

**Comment by sergefdrv** [comment]
I suggest to think of whole distributed systems as something **purposeful**, **sufficient** and **complete**.

In a _functional_ model, a distributed system must be a part of a greater system, playing a certain _role_ within the encompassing system, performing a certain _function_ therein, i.e. have a _purpose_. From this perspective, at the most abstract level, _everything that is used to fulfill the purpose_ of the distributed system, and nothing more, can be considered internal to it, making it _sufficient_ and _complete_.

**Reply by sergefdrv** [reply]
Of course, it matters how we define the system's purpose. For example, let's consider a [state machine replication](https://en.wikipedia.org/wiki/State_machine_replication) system. We can define the purpose of such system as simply serving client requests according to a specified deterministic state machine in a reliable, fault-tolerant manner. Alternatively, we can define the purpose as totally ordering client requests, triggering execution of those requests in that total order on replicas of the deterministic state machine, and responding to the client requests with the corresponding execution results. The former is a _higher-level_ purpose, and the replicated state machine (including, e.g., the database to persistently store its state) would be internal to the correspondingly modeled distributed system; whereas, in the latter case, the replicated state machine would be external w.r.t. the distributed system. So the system's purpose can be defined at different levels of abstraction, and this determines the scope of the modeled system in terms of sufficiency and completeness.

**Reply by anchalshivank** [reply]
I initially approached the distinction between internal and external components from a control-based perspective. My thinking was that internal components are those we can design and manage, while external ones are outside our control.

However, I now see that we can also frame this distinction in terms of the system’s purpose. The purpose of the system can serve as a guiding principle for determining what is internal or external. Components that directly contribute to fulfilling the system's purpose would be considered internal, while those that interact with the system without directly serving that purpose could be treated as external.

That said, how do we define the purpose in a way that remains useful, given that the purpose will change based on the specific use case?

**Reply by sergefdrv** [reply]
The purpose of a distributed system will naturally depend on the purpose and structure of the greater system that it is considered a part of. I think, to make it more useful, one can try to define the purposes of a given type of distributed systems in a way that makes it reusable in the contexts of possibly many other encompassing systems, i.e. by making to more generic.

**Reply by sergefdrv** [reply]
Abstractly speaking, the purpose of any system is to reduce uncertainty, i.e. non-determinism of the external world, in a certain way.

**Comment by sergefdrv** [comment]
Time is closely related to the notion of _change_. Intuitively, we can start by considering time as a _dimension of change_, a continuum for representing the dynamics of the system. Then a dynamically changing system can be represented with a _function of time_, where the image of the system at each point in time corresponds to the _state_ of the system at that moment. 

Assuming there are countably many system states, the time-function effectively defines a sequence of system states or, equally, a _sequence of state transitions_. Time makes each state transition _distinct_ from any other state transition in the time-sequence, even if the corresponding changes in the system state are equivalent. So we can think of time as _ordering of changes_. 

We can take a part of the system, a subsystem, and obtain that subsystem's time-function by projecting the result of the original time-function onto the corresponding partial system state. Partitioning the original time-sequence into contiguous intervals according to the partial state projection, we can relate the subsystem's state transitions to the state transitions of the original system corresponding to the transitions between those intervals. This will preserve the ordering according to the precedence relation.

For any pair of state transitions in the same time-sequence we can tell which one precedes the other in time, thus defining a _precedence relation_. If we take two arbitrary subsystems and try to relate their state transitions in terms of precedence according to the encompassing system, there might be some state transitions that correspond to the same state transition of the encompassing system. We can say that such state transitions are _simultaneous_. The precedence relation should be undefined for simultaneous state transitions and thus become a _partial_ relation.

Time is also closely related to the notion of _causality_. We can consider state transitions as _events_, where one event may _causally depend_ on another event. An event casually depends on another event if a modification of the latter can affect the former. Casual dependency must be an asymmetric and transitive relation. There might be events such that are neither casually depends on the other, so the casual dependency relation can be _partial_, as well.

An event can only casually depend on an event that precedes it in time. Time must ensure _casual consistency_ by ordering events such that this rule is obeyed. 

So we can think of time as _causally consistent partial ordering of changes_.

**Reply by dtornow** [reply]
I believe "simultaneous" implies "at the same point in physical time" whereas the term "concurrent" implies "unrelated". So if two transitions are concurrent they can happen in any order or simultaneously. (I am not an English native speaker tho)

**Reply by sergefdrv** [reply]
> "simultaneous" implies "at the same point in physical time"

which is kinda what's implied by "correspond to the same state transition of the encompassing system". As I put it there, simultaneous transitions are just different projections of the same transition. However, this changes when we think about causality and relax the ordering from total to partial.

> whereas the term "concurrent" implies "unrelated"

indeed, we can say that transitions that don't causally depend on each are _concurrent_. It should not matter how we order concurrent transitions.

**Reply by sergefdrv** [reply]
We should remember that the distributed system that we model must be a part, a subsystem of a greater system. And we should think of its relations to any other subsystem of that greater system in a similar way as we think about relations between subsystems within our distributed system.

**Reply by sergefdrv** [reply]
Some ideas from [Are We There Yet](https://www.youtube.com/watch?v=ScEPu1cs4l0) by Rich Hickey:
- future as a function of the past
- identities associated to series of causally-related values
- state as a mapping from identities to values

By _values_, I believe, Rich means a piece of information. He emphasizes that values are immutable. In our thinking, we abstract a series of causally-related values as an object by associating an _identity_ to the series of values. The system _state_ is an immutable snapshot of the system that is a mapping from object identities to the corresponding values. When there is a _change_, the future is created from the past by applying a function to the state of the past.

**Comment by sergefdrv** [comment]
Time and non-determinism are closely related. If a system is fully determined by the initial assumptions then it's absolutely static, and time is irrelevant to it. Distributed systems are not like that.

Non-determinism primarily manifests itself as uncertainty of changes happening in the system over time. On the other hand, deterministic intermediate transformations within the system should be irrelevant in terms of time. Building upon [previous considerations](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10639812), we can think of _time as causally consistent ordering of non-deterministic changes_.

The ultimate source of non-determinism must be outside of the system because a system is determined by everything that is a part of it. Thus we can think of _non-determinism as external dependencies_ of the system.

If we take a _certain_ state transition of a system then everything it depends on must also be certain. From that perspective, the past is fully deterministic; thus non-determinism belongs to the future. We can say that a dynamic system _consumes (resolves?) non-determinism from outside as it progresses in time_.

**Reply by sergefdrv** [reply]
Although the past is fully deterministic, as a _finite_ system progresses in time, _the system can only retain and expose a bounded amount of determinism_.

**Comment by sergefdrv** [comment]
Building upon previous considerations about [time](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10639812) and [non-determinism](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10661687), we can think of _system state as consistent image of the system or information that is necessary and sufficient for preserving the system's causal consistency_ as it progresses in time. In other words, the system state represents the past of the system, the result of causally consistent partial ordering of changes that possible future changes may causally depend on.

The system state does not, and often cannot, represent the history of the system in full detail; instead, it contains just enough information about the past to preserve the consistency of the system in future, thus connecting the past and the future. Since the system state may lose unnecessary information about the past, and it must do this if the lifetime of the system is indefinite, it may be uncertain about the exact prior history of the system, as if partially retaining the non-determinism absorbed from outside in the past.

**Comment by sergefdrv** [comment]
Interaction is a dynamic process, something that happens over time, and thus it inherently involves non-determinism (see previous considerations about [time](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10639812) and [non-determinism](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10661687)).

We can think of _interaction as causal dependencies between the interacting systems_. In unidirectional interaction only one system depends on the other; in bidirectional interaction both systems depend on each other. A subsystem interacts with the encompassing system by interacting with other subsystems within that encompassing system.

The system that acts in interaction as an external dependency of the other represents a source of non-determinism for the latter. If we consider the interacting systems as one then we shall see that the non-determinism of their interaction must ultimately originate from the outside of both systems. Thus we can also think of _interaction as redistribution of non-determinism_ from outside.

**Reply by sergefdrv** [reply]
Note that the presence of causal dependencies between interacting systems implies that they must also share common state (see [previous considerations about state](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10729558)).

**Comment by sergefdrv** [comment]
A correct system must fulfill its purpose given the assumptions hold. However, when the assumptions are violated, the system may fail to fulfill its purpose. In other words, a system as a whole may fail if the amount of uncertainty (non-determinism) imposed on the system exceeds what it was supposed to handle. In that case, the system will leak the excessive non-determinism imposed on it.

**Comment by sergefdrv** [comment]
In computer science, there are metaphoric concepts of "angelic" and "demonic" non-determinism. Angelic non-determinism aims to favor desired results, whereas demonic non-determinism allows for unpredictable and potentially unfavorable outcomes. We might also want to distinguish those types of non-determinism in our system model.

**Comment by sergefdrv** [comment]
As [noted earlier](https://github.com/orgs/replica-io/discussions/49#discussioncomment-10738392), we can think of _interaction between (sub)systems as establishing new causal relationships_ between events (changes) in those systems or, more precisely, occurrence of events in one system that causally depend on some events from the other system.

We can also think of such _interaction as communicating information about causal dependencies between past events_. The information thus communicated can be of different fidelity, i.e. preserving different amount of details about transitive causal dependencies between past events. The amount of detail communicated determines what causal dependencies the new event can establish with the past events. The amount of information that can be communicated in interaction is limited; therefore, new causal dependencies may only refer directly to a limited part of the actual causal dependency graph.

At the same time, _each act of interaction effectively resolves some amount of non-determinism_ imposed on the recipient system from the outside by determining which from all possible communication events occurs. On the other hand, since only limited amount of information can be communicated, some _uncertainty, i.e. non-determinism, may be re-introduced in communication_ from the perspective of the system where it originates from.

**Comment by sergefdrv** [comment]
We cannot allow absolutely arbitrary interaction, i.e. establishing new casual relationships, between (sub)systems, or we end up with chaos instead of a purposeful system. Therefore, there must be something that determines which interactions are possible and which are not.

Absolutely isolated, unrelated systems should not be able to interact; in order for interaction to happen between systems, they must already have some relationship. Thus _interactions are enabled by already established relationships_.

We can think of those prior relationships enabling interactions as _links or connections_ between systems. Such connections can be established dynamically as a result of _prior interaction_, but ultimately, there must be some set of relationships determined by the way the system was _initially prepared_. Those _initial relationships effectively characterize, define the system_ and its possible behavior.

**Reply by sergefdrv** [reply]
Though, if this is all then there is a problem: once a potential interaction is enabled, there is no way back, it would remain enabled, and nothing would prevent it from happening. Non-trivial distributed, concurrent systems should allow interacting with them in different ways, and prior interactions should influence the future behavior of the system. If possible interactions only ever get enabled but never disabled then the number of possible interactions would monotonically increase as the system keeps interacting. Intuitively, this is not what can happen _without a bound_ in a real system. Therefore, interactions should not only be able to introduce new possible future interactions but also _exclude_ previously enabled interactions.

**Reply by TomasMikula** [reply]
As you know, this is a non-problem in Libretto, as the interaction channels are single-shot, and repeated interaction is achieved via recursive channels, such as a stream, which, btw., can also be "closed" uni-laterally. ("Closed" in quotes, because it's just a regular interaction: choosing the branch that does not recurse further.)

**Comment by sergefdrv** [comment]
Imagine a (possibly infinite) _closed, isolated universe_. Nothing can be added to or removed from the universe, nothing can interfere with it. We can think of the universe as absolutely _static_ and having _total knowledge_ of everything. Now consider a part of the universe as an _open system_ existing within that universe. The boundary of the system separates it from its _environment_, which is the rest of the universe. The system as a whole has _complete knowledge of itself but absolute uncertainty about its environment_, and the other way around.

The system may acquire some new information and so _expand_ its boundary; it may also discard some information that was part of it and so _contract_ its boundary. Because nothing can be added to or removed from the universe, _the total information must be conserved_. So whatever the system learns or forgets must come from and return back to its environment.

Thus we can think of _evolution of an open system as a process of acquiring information from and returning information back to its environment_. Note that the whole process is in principle _reversible_. Also note that the process of just adding or discarding information alone is _monotonic_ and enjoys the _confluence_ property.

The _state_ of the system can be defined as the information the system was initially prepared with plus the information acquired minus the information discarded by the system to the point. Real systems have physical limitations and therefore cannot hold unbounded amounts of information. On the other hand, the system should not discard any information that is still required to fulfill its purpose. So there must be some constraints specifying _what information the system must acquire before it is allowed to discard some specific information and what information it must discard before it is allowed to acquire some new information_. In simple words: the system may be required to forget in order to learn, and learn in order to forget.

Some information that the system acquires from its environment may come from other systems of interest (e.g. input), whereas some information may come from ambient sources (e.g. timing, randomness). Some information that the system holds may be acquired by other systems of interest (e.g. output), whereas some information may be simply dissipated into the ambient environment. What if we define distributed and concurrent systems and their components with acquiring and discarding of information made explicit, so that system's evolution is reversible? How much can we benefit from the associated monotonicity and confluence properties?

**Reply by sergefdrv** [reply]
If we consider evolution of a _finite_ system during a _finite_ time span embedded into a _large enough_ environment, so that we can neglect its interaction with the rest of the universe, we may think of the system combined with the environment as a kind of closed universe. If information is conserved in a closed system then its evolution must be _reversible_, i.e. we can run it backward as well as forward. This mental picture could be useful for formal verification (where one often has to consider all possible executions), deterministic simulation, reproducible testing, reverse-debugging, etc.

**Reply by sergefdrv** [reply]
_Tying all the loose ends of acquiring and discarding information_ (treating it as linear resource?) could help structuring and reasoning about complex systems.

**Reply by TomasMikula** [reply]
Could you please elaborate in what sense is the process of adding or discarding information monotonic and what confluence property it enjoys?

**Reply by TomasMikula** [reply]
> If information is conserved in a closed system then its evolution must be reversible, i.e. we can run it backward as well as forward. This mental picture could be useful for formal verification (where one often has to consider all possible executions), deterministic simulation, reproducible testing, reverse-debugging, etc.

I'm having trouble to see how reversibility would help these (except for reverse-debugging). Do you have a more detailed idea about the others?

**Reply by sergefdrv** [reply]
> in what sense is the process of adding or discarding information monotonic and what confluence property it enjoys?

If we consider just adding information to a system alone then it is monotonic since the volume of information absorbed by the system can only grow. Symmetrically, to see that just discarding information from the system alone is monotonic, observe that the volume of information discarded monotonically grows. Of course, adding and discarding combined together may not be monotonic. 

This may be tricky in a scenario where a system re-acquires some information that it has previously discarded. To overcome this, we may require that any information being acquired by the system must be unique and distinct form any information it has discarded. Or we may consider the volume of acquiring/discarding information as a kind of multi-set.

Confluence follows from monotonicity if we think of adding information as a commutative and associative operation. So if we consider two "points of interaction on the surface" of the system at which the system can absorb information from the environment (at which it can "grow") then it should not matter in which order the system acquires information at those points, i.e. acquiring information through different points is truly concurrent.

**Reply by sergefdrv** [reply]
> > This mental picture could be useful for formal verification (where one often has to consider all possible executions), deterministic simulation, reproducible testing, reverse-debugging, etc.
> 
> I'm having trouble to see how reversibility would help these (except for reverse-debugging).

It was a speculative statement ("could"). However, making inflow of all information explicit (incl. randomness, timing, non-deterministic choices, etc.) can help with deterministic simulation and reproducible testing (making outflow of information explicit is not required for those features). How to make use of this in formal verification is not clear yet, but worth keeping an eye on.

**Reply by TomasMikula** [reply]
> Confluence follows from monotonicity if we think of adding information as a commutative and associative operation. So if we consider two "points of interaction on the surface" of the system at which the system can absorb information from the environment (at which it can "grow") then it should not matter in which order the system acquires information at those points, i.e. acquiring information through different points is truly concurrent.

The order wouldn't matter if the system had no way to observe which came first. However, it is _desirable_ to have the ability to observe which of two inputs arrived first and transition to different states based on the result (that's what _racing_ is for in Libretto).

Here's an example: After clicking "Send" in Gmail, you have a few seconds to "Cancel". So there are two inputs now: the timer and the Cancel button. Depending on which fires first, Gmail either sends the email or cancels sending. Confluence is undesirable.

**Reply by TomasMikula** [reply]
> However, making inflow of all information explicit (incl. randomness, timing, non-deterministic choices, etc.) can help with deterministic simulation and reproducible testing (making outflow of information explicit is not required for those features).

So what you need is making all inputs explicit, not reversibility :)

**Reply by sergefdrv** [reply]
> it is _desirable_ to have the ability to observe which of two inputs arrived first and transition to different states based on the result

That's a good point! I just pondered on this yesterday. Let's first consider how we can implement deterministic, reversible racing and then see how it would work when we really need non-determinism (like in the example you provided).

We can implement racing with a primitive component similar to the [controlled-swap gate](https://en.wikipedia.org/wiki/Fredkin_gate) but generalized to swapping arbitrary values (of the same type) instead of just bits. It would swap the 2nd and 3rd inputs (targets) depending on the value of the 1st input (control bit). So It would have three inputs and send them to the outputs as follows:

$$
\begin{align}
(0, t_1, t_2) \rightarrow (0, t_1, t_2) \\
(1, t_1, t_2) \rightarrow (1, t_2, t_1)
\end{align}
$$

Then we can use the control bit (1st argument) to select which of the two other inputs should "win the race". The "winner" and "looser" will be the 2nd and 3rd outputs, respectively; the control bit will go through as is.

The trick is that, in normal operation, the value of the control bit will be determined at spot, depending on which of the other inputs becomes actually available first. However, in simulation mode (e.g. during testing or verification), we could make the decision deterministically (and thus reproducible) by setting the value of the control bit from outside (or making it even a kind of superposition). We could also record the outputs that would be normally discarded and so be able to backtrack the execution with a kind of wayback machine.

> So what you need is making all inputs explicit, not reversibility :)

As I mentioned, reversibility is good for reverse-debugging, and it may have other uses (e.g. for backtracking when verifying correctness). We may eventually decide to give up on reversibility, in some cases maybe, but it has already given some interesting insights, so I would rather keep it and see where it brings us.

**Reply by michaeljklein** [reply]
Hi, @aakoshh mentioned this to me so here are some initial thoughts: 👋

`The system as a whole has complete knowledge of itself but absolute uncertainty about its environment..`
- By "complete knowledge about itself," do you mean "knowing everything," i.e. solving the halting problem for any included formula, or "fully formally specified"?

`So whatever the system learns or forgets must come from and return back to its environment.`
- I.e. a formal system of resource management? :) 

`So there must be some constraints specifying what information the system must acquire before it is allowed to discard...required to forget in order to learn, and learn in order to forget.`
- I considered similar issues for a while, but instead of solving them directly, my approach is to assign weights or values to information in a real-world-value-relevant way so that the final decision could be as easy as: "I keep any information required to function and maximize total information value."

`Some information that the system acquires from its environment may come from other systems of interest (e.g. input), whereas some information may come from ambient sources (e.g. timing, randomness).`
- I think of this along the axis of "trusted" (e.g. signed, deterministic) data vs. "fuzzy" (e.g. user-recorded sensor, non-deterministic) data.

A general comment r.e. reversible computing:
- A deterministic state machine + a history (e.g. using `gdb` or a blockchain) is more or less reversible so:
  1. Reversibility can, in some sense, be "added on top" of another language
  2. One of the hardest parts to utilizing reversibility is its efficiency. E.g. consider the expected overhead from adding naive reversibility to a SIMD-optimized loop.

**Reply by sergefdrv** [reply]
>> The system as a whole has complete knowledge of itself but absolute uncertainty about its environment..
> 
> By "complete knowledge about itself," do you mean "knowing everything," i.e. solving the halting problem for any included formula, or "fully formally specified"?

It was more of a definition: a system as a part of the universe is all there is to know about a certain part of the universe, and the environment is everything else. The only thing, perhaps, that the system knows about its environment and vise versa is that there is a boundary in-between. So, I think, "knowing everything" is a correct interpretation of that. Not sure how to interpret the halting problem in this context, but if it applies, the system would also know if some information within it is well founded, i.e. "terminates", though it may not be possible to extract this knowledge out of the system into the environment.

>> So whatever the system learns or forgets must come from and return back to its environment.
> 
> I.e. a formal system of resource management? :)

Yes, information is treated as a linear, extensive resource here.

> assign weights or values to information in a real-world-value-relevant way

Do you mean that the weight/value of information is derived from the piece of information by applying some universal function, i.e. it is already present there, or is it some additional information that you keep alongside? 

> so that the final decision could be as easy as: "I keep any information required to function and maximize total information value."

How do we know which information is "required to function"? What happens once the total information reaches a maximal value?

> One of the hardest parts to utilizing reversibility is its efficiency. E.g. consider the expected overhead from adding naive reversibility to a SIMD-optimized loop.

We may choose to optimize reversibility out in normal execution, but keep it for testing/debugging/verifying. We can make it optional at the expense of some extra features, if it becomes too problematic in certain cases. We can also choose the granularity of components. In general, the amount of output "garbage" data required to preserve reversibility cannot exceed the amount of input data, including the ancillary/temporary data.

**Reply by sergefdrv** [reply]
This thread started very abstract and even philosophical. Note that there is a [follow-up thread](https://github.com/orgs/replica-io/discussions/49#discussioncomment-12436415), which is a bit more concrete.

**Comment by sergefdrv** [comment]
Let's try to make the idea described in [the previous comment](https://github.com/orgs/replica-io/discussions/49#discussioncomment-12273776) a bit more concrete. We can model concurrent systems made out of components that are _deterministic_ machines, each with a number of data inputs and the _same_ number of outputs. The communication of a component with the environment happens _only_ by passing _plain_ data items through those inputs/outputs. Each input has a _complementary_ output of the _same type_, so the data items entering the component through the input and exiting through the complementary output have the _same size_ in binary representation.

Inputs and outputs are in general _concurrent_, i.e. data items can enter and exit the component through individual inputs/outputs at any time, unless there is an internal data dependency required by the component's logic. The output data must only depend on the input data and nothing else. In particular, it does not matter in which order data items are supplied/returned through _different_ inputs/outputs, only the content matters, although the order of data items supplied/returned through the same input/output matters.

As data items pass through the component, they would exchange information according to the component's logic, thus creating causal dependencies. Note that the type and the size of data items remains unchanged. Essentially, it's the same piece of memory that enters the component through the input and exits it through the complementary output, but the content can change and/or affect the other inputs/outputs. Causal dependencies may require several inputs to become available and remain within the component until the information exchange between data items can happen, before the can exit the component. We may also require that the whole process is _reversible_, i.e. that we could supply the output data back to the component in the reverse direction in order to reconstruct the input. An example of such component, a generalized controlled-swap gate, was described in [this comment](https://github.com/orgs/replica-io/discussions/49#discussioncomment-12425172).

Multiple components can be combined together by connecting intputs to outputs of the same data type. Each input/output can be connected at most once. The inputs/outputs that remain unconnected constitute the interface of the combined system.

As suggested by @pompon0 in a private discussion, one can think of this model as a collection of deterministic multi-tape multi-head Turing machines with the restriction that tapes are allowed to move only in one direction.

**Reply by sergefdrv** [reply]
Since components in this model are deterministic, _time_ manifests itself primarily in _causal dependencies_ between inputs/outputs. Time can also come up as input in form of a _timestamp_; in that case, there might be some assumptions about the values of timestamps, e.g. that timestamps incoming over the same input line are monotonically increasing and consistent with other inputs.

**Reply by sergefdrv** [reply]
In order to function correctly, components would _require_ some assumptions about their inputs to hold; in return, they'll provide certain _guarantees_ about the outputs. We can think of a _component failure as a violation of its guarantees_. It can be a fail-stop failure, where the faulty component simply stops producing any output, or Byzantine failure, where the faulty component may produce arbitrary output violating some or all of its guarantees. When we connect components together, only if the _provided guarantees are stronger than the requirements_, the composite system could _tolerate_ failure of some its components.

**Reply by sergefdrv** [reply]
NB. This idea was further developed in a blog post: https://replica-io.dev/blog/2025/05/30/a-sketch-of-reversible-deterministic-concurrency-for-distributed-protocols

**Comment by sergefdrv** [comment]
Now that we have [some ideas at the most abstract level](https://github.com/orgs/replica-io/discussions/49?sort=old#discussioncomment-12501075), we should consider modeling of individual nodes within a distributed system and their interaction. How should we think about distributed nodes and means of communication between them?

**Comment by sergefdrv** [comment]
Physically, a distributed system is a collection of largely independent computing machines, called _nodes_, that only interact by communicating data over some kind of network. Generally, communication over the network is not reliable, i.e. data packets may be dropped, delivered multiple times or out of order. Therefore, there are usually some higher-level mechanisms built on top of the unreliable low-level network in order to ensure more reliable communication, e.g. ordered and reliable (under certain assumptions) delivery of data without duplication. Communication between nodes generally happens concurrently and asynchronously; therefore, nodes and the communication system connecting them have to deal with concurrency. So we can think of _individual nodes as concurrent systems interacting with a common communication system, which is also concurrent_.

On the other hand, logically, one may want to distinguish pieces of functionality within a distributed system that span multiple physical nodes. In fact, the communication system is an example of this. So we can conceptually think of the overall _distributed system as a collection of interacting concurrent systems possibly spanning multiple physical nodes_.

Different nodes may play different _roles_ depending on their function withing the distributed (sub-)system, e.g. client, leader, follower, etc. A single node may even play multiple roles at once or dynamically change its role (de-/activate different roles).

**Reply by sergefdrv** [reply]
One can conceive different mechanisms of communication between distributed nodes: unordered or preserving the order of messages, with best-effort or with (bounded) reliable delivery guarantees, point-to-point or disseminating data among many nodes, etc. However, we can always design communication components such that there is an _illusion of local interaction with a remote/distributed component through its local representation_.

**Reply by sergefdrv** [reply]
Note that conceptual distributed components spanning multiple physical nodes can be _projected_ onto individual nodes and thereby manifest _as their local representations_.

**Reply by sergefdrv** [reply]
_Any mechanism of communication_ between distributed nodes is necessarily distributed itself and _can be conceptually thought of as a distributed component spanning across multiple nodes_.

**Reply by sergefdrv** [reply]
Perhaps, the basic communication mechanism should be a point-to-point link/channel. It may have different properties, but we should probably prefer channels with bounded capacity that preserve message ordering without dropping.

**Comment by sergefdrv** [comment]
We can adopt a _uniform approach to modeling of distributed and local concurrent interaction_, in which the difference between local (intra-node) and remote (inter-node, distributed) interaction is not treated as fundamental but emergent from the properties of communication components.

**Comment by sergefdrv** [comment]
_Dynamism and adaptivity_ can be expressed _in a principally static description by capturing regular patterns as potential structure_ (i.e. any shape that the system might take), whereas the actual structure representing _emergent dynamics_ is determined _through activation_ (e.g. data-dependent) _of a bounded subset of the potential structure_ under certain conditions.

**Reply by sergefdrv** [reply]
We could also consider higher-order structures (i.e. structures operating on structures as first-class entities), but that could significantly increase complexity of reasoning.

**Comment by sergefdrv** [comment]
We can think of _coordination as correlation_ (entanglement) and _fault tolerance as coarse-graining with error correction_, both _manifesting in redundant symmetry_, whereby _failures_ can be thought of _as symmetry-breaking noise_.

**Reply by sergefdrv** [reply]
The system's finite information capacity is thus shared between redundancy (fault tolerance) and useful coordination, so greater resilience to failures necessarily comes at the cost of coordination capacity.
